File: benchmarks/calc_pi.js
const WARMUP_RUNS = 3
const WARMUP_N = 10000;


function calcPi(nTerms) {
    let pi = 0.0;

    for(let i = 0; i < nTerms; i++) {
        let sign = 1;

        if(i % 2 == 1)
            sign = -1;
    
        pi += sign / (2.0 * i + 1);
    }
    
    return 4 * pi;
}

function runBenchmark(nTerms) {
    for(let i =0; i < WARMUP_RUNS; i++)
        calcPi(WARMUP_N);

    // Start benchmark timings
    const wallStart = Date.now();
    const cpuStart = performance.now();

    const pi = calcPi(nTerms); // Run benchmark

    // End benchmark timing
    const wallEnd = Date.now();
    const cpuEnd = performance.now();

    // Calculate Results in seconds
    const wallTimeSec = (wallEnd - wallStart) / 1000;
    const cpuTimeSec = (cpuEnd - cpuStart) / 1000;

    // Output results
    console.log(`${wallTimeSec} ${cpuTimeSec} ${pi}`)
}

function main() {
    const args = process.argv.slice(2);  // ignore 'node' and program name in CL args
    
    if(args.length < 1) { // No nTerms provided
        console.error('Please provide the number of terms as a command-line argument.');
        process.exit(1);
    }

    const nTerms = Number(args[0]); // Parse nTerms to numerical value

    // nTerms arg is not valid integer
    if (isNaN(nTerms) || nTerms <= 0 || !Number.isInteger(nTerms)) {
        console.error(`Invalid number format: ${args[0]}`);
        process.exit(1);
    }

    runBenchmark(nTerms); // Run the benchmark
}

// Run main if this script is executed directly
if (require.main === module) {
    main();
}

File: benchmarks/calc_pi.py
import time
import sys

def calc_pi(n_terms: int) -> float:
    pi: float = 0.0

    for i in range(n_terms):
        sign: int = 1

        if i % 2 == 1:
            sign = -1

        pi += sign / (2.0 * i + 1)
    
    return 4 * pi

def run_benchmark(n_terms: int) -> None:
    for i in range(10): # Warmup Python environment
        calc_pi(10000)

    # Start timings
    start_wall: float = time.time()
    start_cpu: float = time.process_time()

    pi = calc_pi(n_terms) # Run benchmark

    # End timings
    end_cpu: float = time.process_time()
    end_wall: float = time.time()

    # Output results
    print(f"{end_wall - start_wall:.6f} {end_cpu - start_cpu:.6f} {pi}")

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Please provide the number of terms as a command-line argument.")
        sys.exit(1)

    try:
        n_terms = int(sys.argv[1])
    except ValueError:
        print(f"Invalid number format: {sys.argv[1]}")
        sys.exit(1)

    run_benchmark(n_terms)